processing/src/main/java/org/apache/carbondata/processing/newflow/steps/writer/DataWriterProcessorStepImpl.java

```diff
@@ -0,0 +1,360 @@
+    isUseInvertedIndex =
+        CarbonDataProcessorUtil.getIsUseInvertedIndex(configuration.getDataFields());
+
+    if (!(new File(storeLocation).exists())) {
+      LOGGER.error("Local data load folder location does not exist: " + storeLocation);
+      return;
+    }
+
+    int[] dimLensWithComplex =
+        (int[]) configuration.getDataLoadProperty(DataLoadProcessorConstants.DIMENSION_LENGTHS);
+    List<Integer> dimsLenList = new ArrayList<Integer>();
+    for (int eachDimLen : dimLensWithComplex) {
+      if (eachDimLen != 0) dimsLenList.add(eachDimLen);
+    }
+    dimLens = new int[dimsLenList.size()];
+    for (int i = 0; i < dimsLenList.size(); i++) {
+      dimLens[i] = dimsLenList.get(i);
+    }
+
+    this.dimensionCount = configuration.getDimensionCount();
+    this.noDictionaryCount = configuration.getNoDictionaryCount();
+    this.complexDimensionCount = configuration.getComplexDimensionCount();
+    this.measureCount = configuration.getMeasureCount();
+
+    int simpleDimsCount = this.dimensionCount - complexDimensionCount;
+    int[] simpleDimsLen = new int[simpleDimsCount];
+    for (int i = 0; i < simpleDimsCount; i++) {
+      simpleDimsLen[i] = dimLens[i];
+    }
+
+    CarbonTable carbonTable = CarbonMetadata.getInstance().getCarbonTable(
+        tableIdentifier.getDatabaseName() + CarbonCommonConstants.UNDERSCORE + tableIdentifier
+            .getTableName());
+    wrapperColumnSchema = CarbonUtil
+        .getColumnSchemaList(carbonTable.getDimensionByTableName(tableIdentifier.getTableName()),
+            carbonTable.getMeasureByTableName(tableIdentifier.getTableName()));
+    colCardinality = CarbonUtil.getFormattedCardinality(dimLensWithComplex, wrapperColumnSchema);
+    segmentProperties = new SegmentProperties(wrapperColumnSchema, colCardinality);
+    // Actual primitive dimension used to generate start & end key
+
+    keyGenerator = segmentProperties.getDimensionKeyGenerator();
+
+    //To Set MDKey Index of each primitive type in complex type
+    int surrIndex = simpleDimsCount;
+    Iterator<Map.Entry<String, GenericDataType>> complexMap =
+        CarbonDataProcessorUtil.getComplexTypesMap(configuration.getDataFields()).entrySet()
+            .iterator();
+    complexIndexMap = new HashMap<Integer, GenericDataType>(complexDimensionCount);
+    while (complexMap.hasNext()) {
+      Map.Entry<String, GenericDataType> complexDataType = complexMap.next();
+      complexDataType.getValue().setOutputArrayIndex(0);
+      complexIndexMap.put(simpleDimsCount, complexDataType.getValue());
+      simpleDimsCount++;
+      List<GenericDataType> primitiveTypes = new ArrayList<GenericDataType>();
+      complexDataType.getValue().getAllPrimitiveChildren(primitiveTypes);
+      for (GenericDataType eachPrimitive : primitiveTypes) {
+        eachPrimitive.setSurrogateIndex(surrIndex++);
+      }
+    }
+  }
+
+  private void initDataHandler() {
```

Comment:
can we move all these initialization logic into CarbonFactHandlerFactory? I think the handlerModel is created from configuration, so passing configuration to the factory can create both model and the handler.